src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp

```diff
@@ -623,30 +618,37 @@ void ShenandoahDirectCardMarkRememberedSet::merge_write_table(HeapWord* start, s
   for (size_t i = 0; i < num; i++) {
     read_table[i] &= write_table[i];
   }
+
+  log_info(gc, remset)("Finished merging write_table into read_table.");
 }
 
-// Destructively copy the write table to the read table, and clean the write table.
-void ShenandoahDirectCardMarkRememberedSet::reset_remset(HeapWord* start, size_t word_count) {
-  size_t start_index = card_index_for_addr(start);
+void ShenandoahDirectCardMarkRememberedSet::swap_card_tables() {
+  CardTable::CardValue* new_ptr = _card_table->swap_read_and_write_tables();
+
 #ifdef ASSERT
-  // avoid querying card_index_for_addr() for an address past end of heap
-  size_t end_index = card_index_for_addr(start + word_count - 1) + 1;
+  CardValue* start_bp = &(_card_table->write_byte_map())[0];
+  CardValue* end_bp = &(new_ptr)[_card_table->last_valid_index()];
+
+  while (start_bp < end_bp) {
```

Comment:
Same issue here as mentioned above.  So our validation test doesn't catch our failure to overwrite last index value with clean value.